tests/lib/linter/linter.js

```diff
@@ -7590,6 +7590,44 @@ var a = "test2";
                 linter.verify("0", { rules: { "test-rule": "error" } });
             }, /Fixable rules must set the `meta\.fixable` property/u);
         });
+
+        it("should stop fixing if a circular fix is detected", () => {
+            linter.defineRule("circular-fix", {
+                meta: {
+                    fixable: "whitespace"
+                },
+                create(context) {
+                    return {
+                        Program(node) {
+
+                            // Alternate between adding and removing a hyphen.
+                            const sourceCode = context.getSourceCode();
+                            const hasLeadingHyphen = sourceCode.getText(node).startsWith("-");
+
+                            context.report({
+                                node,
+                                message: "Circular fix detected",
+                                fix: fixer => (hasLeadingHyphen
+                                    ? fixer.removeRange([0, 1])
+                                    : fixer.insertTextBefore(node, "-"))
+                            });
+                        }
+                    };
+                }
+            });
+
+            const initialCode = "-a";
+            const fixResult = linter.verifyAndFix(initialCode, { rules: { "circular-fix": "error" } });
+
+            assert.strictEqual(fixResult.fixed, true, "Fixing should have been applied but stopped due to circular fixes.");
+            assert.strictEqual(fixResult.output, "-a", "Output should match the original input due to circular fixes.");
+            assert.strictEqual(fixResult.messages.length, 1, "There should be one remaining lint message after detecting circular fixes.");
+            assert.strictEqual(fixResult.messages[0].message, "Circular fix detected", "Message should match the last fix attempt.");
```

Comment:
I think there should be different messages for adding and removing a hyphen as this doesn't really ensure that the message is correct for the final output.